<a href="https://colab.research.google.com/github/sayarghoshroy/Analysis/blob/main/Statisticial_Tests.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [1]:
# Problems on statistical tests

In [2]:
import numpy as np
import scipy as sc
from scipy.stats import f
from scipy.stats import t

### $ Problem\ 3$

In [3]:
data = [[69, 153],
        [74, 175],
        [68, 155],
        [70, 135],
        [72, 172],
        [67, 150],
        [66, 115],
        [70, 137],
        [76, 200],
        [68, 130],
        [72, 140],
        [79, 265],
        [74, 185],
        [67, 112],
        [66, 140],
        [71, 150],
        [74, 165],
        [75, 185],
        [75, 210],
        [71, 149]]

data = np.asarray(data).T

In [4]:
# Computing Components
alpha = 0.05
n = data.shape[1]
sample_mean = np.mean(data, axis = 1)
target_mean = np.asarray([70, 170])
difference = np.asmatrix(sample_mean - target_mean).T
sample_covariance = np.cov(data)
known_covariance = np.asmatrix([[20, 100], [100, 1000]])

In [5]:
# Computing T-Squared Stat
Z_squared_stat = np.ravel(n * np.matmul(np.matmul(difference.T, np.linalg.inv(known_covariance)), difference))[0]
print('Calculated Z-square statistic value = ' + str(Z_squared_stat))
critical_value = 5.991
print('Chi Squared Distibution Percentage Points for: ')
print('Degrees of Freedom = p = ' + str(data.shape[0]) + ' with alpha = ' + str(alpha)+ ' is ' + str(critical_value))

Calculated Z-square statistic value = 10.260900000000017
Chi Squared Distibution Percentage Points for: 
Degrees of Freedom = p = 2 with alpha = 0.05 is 5.991


$We\ therefore\ reject\ H_{0}\ as\ calculated\ value\ of\ the\ Z^2\ statistic\ =\ 10.261\ is\ \ge\ \chi^2_{\alpha, p}\ =\ 5.991$

## $Problem\ 6$

In [6]:
data = [[51, 36, 50, 35, 42],
        [27, 20, 26, 17, 27],
        [37, 22, 41, 37, 30],
        [42, 36, 32, 34, 27],
        [27, 18, 33, 14, 29],
        [43, 32, 43, 35, 40],
        [41, 22, 36, 25, 38],
        [38, 21, 31, 20, 16],
        [36, 23, 27, 25, 28],
        [26, 31, 31, 32, 36],
        [29, 20, 25, 26, 25]]

data = np.asarray(data).T

n = data.shape[1]
sample_mean = np.mean(data, axis = 1)
sample_covariance = np.cov(data)

target_mean = np.asarray([30, 25, 40, 25, 30])

In [7]:
def T_square_value(alpha, p, v):
  # 'v' being the number of degrees of freedom 
  # 'p' - variable setting

  return f.isf(alpha, p, v - p + 1) / (v - p + 1) * (v * p)

def T_value(alpha, p, v):
  return np.sqrt(T_square_value(alpha, p, v))

def get_t_value(alpha, v):
  return t.isf(alpha, v)

In [8]:
print('Constructing 95 % Simultaneous Confidence intervals:')
print()

critical_T_value = T_value(alpha, data.shape[0], data.shape[1] - 1)

for case in range(0, 5):
  component = np.sqrt(sample_covariance[case, case] / data.shape[1])
  lower_bound = sample_mean[case] - critical_T_value * component
  upper_bound = sample_mean[case] + critical_T_value * component
  print('Range for mu suffix ' + str(case + 1) + ' = (' + str(lower_bound) + ', ' + str(upper_bound) + ')')

Constructing 95 % Simultaneous Confidence intervals:

Range for mu suffix 1 = (21.38216673123666, 50.79965145058153)
Range for mu suffix 2 = (13.170678931334397, 37.9202301595747)
Range for mu suffix 3 = (19.888003712834678, 48.29381446898351)
Range for mu suffix 4 = (12.823053128113768, 41.72240141734078)
Range for mu suffix 5 = (16.81671024417858, 44.63783521036687)


In [9]:
print('Constructing 95 % Bonferroni Confidence intervals:')
print()

critical_t_value = get_t_value(alpha / (2 * data.shape[0]), data.shape[1] - 1)

for case in range(0, 5):
  component = np.sqrt(sample_covariance[case, case] / data.shape[1])
  lower_bound = sample_mean[case] - critical_t_value * component
  upper_bound = sample_mean[case] + critical_t_value * component
  print('Range for mu suffix ' + str(case + 1) + ' = (' + str(lower_bound) + ', ' + str(upper_bound) + ')')

Constructing 95 % Bonferroni Confidence intervals:

Range for mu suffix 1 = (28.38145846944604, 43.80035971237214)
Range for mu suffix 2 = (19.059330904434496, 32.0315781864746)
Range for mu suffix 3 = (26.64658822267579, 41.5352299591424)
Range for mu suffix 4 = (19.699064848701376, 34.84638969675317)
Range for mu suffix 5 = (23.436180670740228, 38.018364783805225)


In [10]:
print('Testing with a Bonferroni Critical Value:')
print()

critical_t_value = get_t_value(alpha / (2 * data.shape[0]), data.shape[1] - 1)
print('Bonferroni Critical Value = ' + str(critical_t_value))
print()

for case in range(0, 5):
  t_value = (sample_mean[case] - target_mean[case]) * np.sqrt(data.shape[1]) / np.sqrt(sample_covariance[case, case])
  print('Absolute value t for case ' + str(case + 1) + ' = ' + str(np.abs(t_value)))
  print('t suffix ' + str(case + 1) + ' is >= Bonferroni Critical Value: ' + str(np.abs(t_value) >= critical_t_value))
  print()

Testing with a Bonferroni Critical Value:

Bonferroni Critical Value = 3.169272667175838

Absolute value t for case 1 = 2.5039075607189867
t suffix 1 is >= Bonferroni Critical Value: False

Absolute value t for case 2 = 0.2665219286236194
t suffix 2 is >= Bonferroni Critical Value: False

Absolute value t for case 3 = 2.5156788157740264
t suffix 3 is >= Bonferroni Critical Value: False

Absolute value t for case 4 = 0.951044821135667
t suffix 4 is >= Bonferroni Critical Value: False

Absolute value t for case 5 = 0.316128991138271
t suffix 5 is >= Bonferroni Critical Value: False



##### $Therefore,\ we\ cannot\ reject\ the\ Null\ Hypothesis.$

In [11]:
# ^_^ Thank You